src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/VirtualProductFileType.php

```diff
@@ -90,7 +90,6 @@ public function __construct(
     public static function getSubscribedEvents(): array
     {
         return [
-            FormEvents::PRE_SET_DATA => 'adaptSelf',
             FormEvents::PRE_SUBMIT => 'adaptSelf',
```

Comment:
@jolelievre 

the `adaptSelf` function would cancel the constraint on the field `name`, because in this case `has_file` is always false... 🤔 

I tested to push the form without virtual product, and it works fine.

What do you think ?

Reply:
Not pretty 🤷‍♀️  however if we need to be able to have empty filename then I guess it's OK

Reply:
well, filename should not be empty, if the virtual product `has_file` toggle is checked.

So the `adaptSelf` function's job was to cancel the constraint if `has_file` was unchecked, `false`.

But `has_file` would always be false on the `PRE_SET_DATA` form event, so the constraint was canceled all the time, that's why I removed it.

Not sure if my explanation is clear :D

Reply:
@matthieu-rolland Instead of using form events to rebuild a form field, what about a custom constraint on the parent form ?
Such constraint should use data from both fields to validate the parent form globally, and add violations to the child field "name" if required.
VirtualProductFileType::configureOptions() could set a global constraint in parent form options :
- https://symfony.com/doc/4.4/reference/constraints/Callback.html
- or your own custom constraint class if it is worth it

Reply:
I used validation group instead, which seems to be perfect for optional validation depending on an other field :+1

Reply:
I used validation group instead, which seems to be perfect for optional validation depending on an other field :+1

Reply:
@matthieu-rolland Very elegant and clever :)